floors/fetcher.go

```diff
@@ -0,0 +1,321 @@
+	glog.Errorf("floor fetcher worker panicked: %v", p)
+}
+
+func NewPriceFloorFetcher(config config.PriceFloors, httpClient *http.Client, metricEngine metrics.MetricsEngine) *PriceFloorFetcher {
+	if !config.Enabled {
+		return nil
+	}
+
+	floorFetcher := PriceFloorFetcher{
+		pool:            pond.New(config.Fetcher.Worker, config.Fetcher.Capacity, pond.PanicHandler(workerPanicHandler)),
+		fetchQueue:      make(FetchQueue, 0, 100),
+		fetchInProgress: make(map[string]bool),
+		configReceiver:  make(chan FetchInfo, config.Fetcher.Capacity),
+		done:            make(chan struct{}),
+		cache:           freecache.NewCache(config.Fetcher.CacheSize * 1024 * 1024),
+		httpClient:      httpClient,
+		time:            &timeutil.RealTime{},
+		metricEngine:    metricEngine,
+	}
+
+	go floorFetcher.Fetcher()
+
+	return &floorFetcher
+}
+
+func (f *PriceFloorFetcher) SetWithExpiry(key string, value json.RawMessage, cacheExpiry int) {
+	f.cache.Set([]byte(key), value, cacheExpiry)
+}
+
+func (f *PriceFloorFetcher) Get(key string) (json.RawMessage, bool) {
+	data, err := f.cache.Get([]byte(key))
+	if err != nil {
+		return nil, false
+	}
+
+	return data, true
+}
+
+func (f *PriceFloorFetcher) Fetch(config config.AccountPriceFloors) (*openrtb_ext.PriceFloorRules, string) {
+	if f == nil || !config.UseDynamicData || len(config.Fetcher.URL) == 0 || !validator.IsURL(config.Fetcher.URL) {
+		return nil, openrtb_ext.FetchNone
+	}
+
+	// Check for floors JSON in cache
+	if result, found := f.Get(config.Fetcher.URL); found {
+		var fetchedFloorData openrtb_ext.PriceFloorRules
+		if err := json.Unmarshal(result, &fetchedFloorData); err != nil || fetchedFloorData.Data == nil {
+			return nil, openrtb_ext.FetchError
+		}
+		return &fetchedFloorData, openrtb_ext.FetchSuccess
+	}
+
+	//miss: push to channel to fetch and return empty response
+	if config.Enabled && config.Fetcher.Enabled && config.Fetcher.Timeout > 0 {
+		fetchInfo := FetchInfo{AccountFloorFetch: config.Fetcher, FetchTime: f.time.Now().Unix(), RefetchRequest: false}
+		f.configReceiver <- fetchInfo
+	}
+
+	return nil, openrtb_ext.FetchInprogress
+}
+
+func (f *PriceFloorFetcher) worker(config config.AccountFloorFetch) {
+	floorData, fetchedMaxAge := f.fetchAndValidate(config)
+	if floorData != nil {
+		// Update cache with new floor rules
+		glog.Infof("Updating Value in cache for URL %s", config.URL)
+		cacheExpiry := config.MaxAge
+		if fetchedMaxAge != 0 && fetchedMaxAge > config.Period && fetchedMaxAge < math.MaxInt32 {
+			cacheExpiry = fetchedMaxAge
+		}
+		floorData, err := json.Marshal(floorData)
+		if err != nil {
+			glog.Errorf("Error while marshaling fetched floor data for url %s", config.URL)
+		} else {
+			f.SetWithExpiry(config.URL, floorData, cacheExpiry)
+		}
+	}
+
+	// Send to refetch channel
+	f.configReceiver <- FetchInfo{AccountFloorFetch: config, FetchTime: f.time.Now().Add(time.Duration(config.Period) * time.Second).Unix(), RefetchRequest: true}
```

Comment:
Suppose the `config.AccountFloorFetch` comes with a mispelled URL. We would expect `f.fetchFloorRulesFromURL(config)` to always return an error in line 224 below:
```
223 func (f *PriceFloorFetcher) fetchAndValidate(config config.AccountFloorFetch) (*openrtb_ext.PriceFloorRules, int) {
224     floorResp, maxAge, err := f.fetchFloorRulesFromURL(config)
225     if err != nil {
226         glog.Errorf("Error while fetching floor data from URL: %s, reason : %s", config.URL, err.Error())
227         return nil, 0
228     }
floors/fetcher.go
```
Do we want to refetch? Should we not discard this faulty client info altogether?
```
152   func (f *PriceFloorFetcher) worker(config config.AccountFloorFetch) {
153 -     floorData, fetchedMaxAge := f.fetchAndValidate(config)
    +     floorData, fetchedMaxAge, err := f.fetchAndValidate(config)
    +     if isReoccurringError(err) {
    +         return
    +     }
154       if floorData != nil {
155           // Update cache with new floor rules
156           glog.Infof("Updating Value in cache for URL %s", config.URL)
157           cacheExpiry := config.MaxAge
158           if fetchedMaxAge != 0 && fetchedMaxAge > config.Period && fetchedMaxAge < math.MaxInt32 {
159               cacheExpiry = fetchedMaxAge
160           }
161           floorData, err := json.Marshal(floorData)
162           if err != nil {
163               glog.Errorf("Error while marshaling fetched floor data for url %s", config.URL)
164           } else {
165               f.SetWithExpiry(config.URL, floorData, cacheExpiry)
166           }
167       }
168  
169       // Send to refetch channel
170       f.configReceiver <- FetchInfo{AccountFloorFetch: config, FetchTime: f.time.Now().Add(time.Duration(config.Period) * time.Second).Unix(), RefetchRequest: true}
171  
172   }
floors/fetcher.go
```
The implementation or `isReoccurringError(err)` should be able to tell apart situations where it's not worth trying to fetch again. Thoughts?

Reply:
If the URL is invalid then we do not pass it to fetcher as we have check in place,`validator.IsURL(config.Fetcher.URL)`

But if your saying the unreachable for long time, then we can have mechanism to discard the URL, but we don't have any way to find out is that url is temporarily down or permanent

We have to keep track of failures

let  me know your thoughts on this

Reply:
What do you think about a max number of retries? Maybe that field could be added to the `FetchInfo` struct and made host-configurable.
```
152 - func (f *PriceFloorFetcher) worker(config config.AccountFloorFetch) {
152 + func (f *PriceFloorFetcher) worker(config config.AccountFloorFetch, hostMaxPriceFloorFetchRetries int) {
153       floorData, fetchedMaxAge := f.fetchAndValidate(config)
154       if floorData != nil {
155           // Update cache with new floor rules
156           glog.Infof("Updating Value in cache for URL %s", config.URL)
157           cacheExpiry := config.MaxAge
158           if fetchedMaxAge != 0 {
159               cacheExpiry = fetchedMaxAge
160           }
161           floorData, err := json.Marshal(floorData)
162           if err != nil {
163               glog.Errorf("Error while marshaling fetched floor data for url %s", config.URL)
164           } else {
165               f.SetWithExpiry(config.URL, floorData, cacheExpiry)
166           }
167       }
168  
169       // Send to refetch channel
    +     if config.retryCount < hostMaxPriceFloorFetchRetries { // horrible name LOL
    +         config.retryCount++
170           f.configReceiver <- FetchInfo{AccountFloorFetch: config, FetchTime: f.time.Now().Add(time.Duration(config.Period) * time.Second).Unix(), RefetchRequest: true}
    +     }
171  
172   }
floors/fetcher.go
```
The reason I believe this is necessary is because `time.Duration(config.Period)` gets added to `time.Now()` in line 170 above making an incorrect URL to throw 404s indefinitely every 5 minutes. What do you think?

Reply:
It seems right to me, to discard the invalid URLs
I will add mechanism to discard long failing URLs

Reply:
Also what should be default value for retries?

Reply:
Default value set to 10